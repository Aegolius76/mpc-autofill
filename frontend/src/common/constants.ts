export const ProjectName = "MPC Autofill";

import { CardType, Faces } from "@/common/types";

export const Card: CardType = "CARD";
export const Cardback: CardType = "CARDBACK";
export const Token: CardType = "TOKEN";

export const CardTypeSeparator = ":";

export const CardTypePrefixes: { [prefix: string]: CardType } = {
  "": Card,
  b: Cardback,
  t: Token,
};

export const ReversedCardTypePrefixes = Object.fromEntries(
  Object.keys(CardTypePrefixes).map((prefix: string) => [
    CardTypePrefixes[prefix],
    prefix.length > 0 ? prefix + CardTypeSeparator : prefix,
  ])
);

export const Front: Faces = "front";
export const Back: Faces = "back";
export const FaceSeparator = "|";

export const ToggleButtonHeight = 38; // pixels

export const MinimumDPI = 0;
export const MaximumDPI = 1500;
export const DPIStep = 50;

export const MaximumSize = 30; // megabytes
export const SizeStep = 1;

export const CSRFKey = "csrftoken";
export const SearchSettingsKey = "searchSettings";
export const GoogleAnalyticsConsentKey = "googleAnalyticsConsent";
export const BackendURLKey = "backendURL";

export const Brackets: Array<number> = [
  18, 36, 55, 72, 90, 108, 126, 144, 162, 180, 198, 216, 234, 396, 504, 612,
];

export const ProjectMaxSize: number = Brackets[Brackets.length - 1];

export enum QueryTags {
  BackendSpecific = "backendSpecific",
}
