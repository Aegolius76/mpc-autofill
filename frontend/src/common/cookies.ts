/**
 * Data API for interacting with anything stored in cookies.
 */

import Cookies from "js-cookie";
import Ajv from "ajv";
import {
  SearchSettings,
  SourceDocument,
  SourceDocuments,
  SourceRow,
} from "./types";
import { searchSettingsSchema } from "./schemas";
import {
  CSRFCookie,
  SearchSettingsCookie,
  GoogleAnalyticsConsentCookie,
  MaximumDPI,
  MaximumSize,
  MinimumDPI,
  BackendURLCookie,
} from "./constants";

const ajv = new Ajv();

//# region CSRF
// TODO: unsure if we still need this.

export function getCSRFHeader(): HeadersInit | undefined {
  const csrfToken = Cookies.get(CSRFCookie);
  if (csrfToken != null) {
    return { "X-CSRFToken": csrfToken };
  }
  return undefined;
}

//# endregion

//# region search settings

export function getCookieSearchSettings(
  sourceDocuments: SourceDocuments
): SearchSettings {
  /**
   * Get search settings from cookie data. If valid data is retrieved,
   * ensure that all `sourceDocuments` are included in the returned settings,
   * with any new sources that weren't previously included added to the end and enabled.
   */

  const rawSettings = JSON.parse(Cookies.get(SearchSettingsCookie) ?? "{}");
  const validate = ajv.compile(searchSettingsSchema);
  const rawSettingsValid = validate(rawSettings);
  if (rawSettingsValid) {
    // reconcile against sourceDocuments
    const sourceInDatabaseSet: Set<number> = new Set(
      Object.values(sourceDocuments).map((sourceDocument) => sourceDocument.pk)
    );
    const sources: Array<SourceRow> = rawSettings.sourceSettings.sources ?? [];
    const sourceInCookieSet: Set<number> = new Set(
      sources.map((row) => row[0])
    );
    // one fat line of reconciliation, good luck reading this future nick! i wrote this at 12:26am.
    rawSettings.sourceSettings.sources = sources
      .filter((row: SourceRow) => sourceInDatabaseSet.has(row[0]))
      .concat(
        Array.from(sourceInDatabaseSet)
          .filter((pk: number) => !sourceInCookieSet.has(pk))
          .map((pk: number) => [pk, true])
      );
    return rawSettings;
  } else {
    // default settings
    return {
      searchTypeSettings: { fuzzySearch: false },
      sourceSettings: {
        sources: Object.values(sourceDocuments).map(
          (sourceDocument: SourceDocument) => [sourceDocument.pk, true]
        ),
      },
      filterSettings: {
        minimumDPI: MinimumDPI,
        maximumDPI: MaximumDPI,
        maximumSize: MaximumSize,
      },
    };
  }
}

export function setCookieSearchSettings(settings: SearchSettings): void {
  Cookies.set(SearchSettingsCookie, JSON.stringify(settings), {
    expires: 365,
    sameSite: "strict",
  });
}

//# endregion

//# region google analytics consent

export function getGoogleAnalyticsConsent(): boolean | undefined {
  const rawConsent = Cookies.get(GoogleAnalyticsConsentCookie);
  return rawConsent != undefined ? JSON.parse(rawConsent) === true : undefined;
}

export function setGoogleAnalyticsConsent(consent: boolean): void {
  Cookies.set(GoogleAnalyticsConsentCookie, JSON.stringify(consent), {
    expires: 365,
    sameSite: "strict",
  });
}

//# endregion

//# region backend

export function getCookieBackendURL(): string | undefined {
  return Cookies.get(BackendURLCookie);
}

export function setCookieBackendURL(url: string): void {
  Cookies.set(BackendURLCookie, url, {
    expires: 365,
    sameSite: "strict",
  });
}

export function clearCookieBackendURL(): void {
  Cookies.remove(BackendURLCookie);
}

//# endregion
