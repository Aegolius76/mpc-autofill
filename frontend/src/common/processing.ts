/**
 * This module contains functions which sanitise and transform user inputs for cards to query into useful formats.
 */

import {
  Card,
  CardTypePrefixes,
  CardTypeSeparator,
  FaceSeparator,
  SelectedImageSeparator,
} from "@/common/constants";
import {
  DFCPairs,
  ProcessedLine,
  ProjectMember,
  SearchQuery,
} from "@/common/types";

export function sanitiseWhitespace(text: string): string {
  /**
   * Clean any instances of doubled-up whitespace from `text`.
   */

  const re = / +(?= )/g;
  return text.replaceAll(re, "").trim();
}

export function stripTextInParentheses(text: string): string {
  /**
   * Remove all text within (parentheses) from `text`.
   * Does not handle (nested (parentheses)). TODO: update this function to do this
   */

  const re = /[([].*?[)\]]/g;
  return sanitiseWhitespace(text.replaceAll(re, ""));
}

export function processQuery(query: string): string {
  /**
   * Process `query` by converting to lowercase, removing all punctuation, and sanitising whitespace.
   * Note that hyphens are not removed due to how Elasticsearch's classic tokenizer works:
   * https://www.elastic.co/guide/en/elasticsearch/reference/7.17/analysis-classic-tokenizer.html
   */

  // TODO: remove any numbers from the front
  // escaping \[ is technically unnecessary, but I think it's more readable to escape it
  return sanitiseWhitespace(
    query
      .toLowerCase()
      .trim()
      // eslint-disable-next-line
      .replace(/[~`!@#$%^&*(){}\[\];:"'<,.>?/\\|_+=]/g, "")
  );
}

export function processPrefix(query: string): SearchQuery {
  /**
   * Identify the prefix of a query. For example, `query`="t:goblin" would yield
   *   {query: "goblin", card_type: TOKEN}.
   */

  for (const [prefix, cardType] of Object.entries(CardTypePrefixes)) {
    if (
      prefix !== "" &&
      query
        .trimStart()
        .toLowerCase()
        .startsWith(`${prefix}${CardTypeSeparator}`)
    ) {
      return {
        query: processQuery(query.trimStart().slice(prefix.length + 1)),
        card_type: cardType,
      };
    }
  }
  return { query: processQuery(query), card_type: CardTypePrefixes[""] };
}

function unpackLine(
  line: string
): [number, [string, string | null] | null, [string, string | null] | null] {
  /**
   * Unpack `line` into its constituents.
   *
   * Inputs to this function are unpacked according to the below schema. For example, consider `4x opt@1234 | char@xyz`:
   *      4x                opt        @        1234         |      char       @        xyz
   * â””â”€ quantity â”€â”€â”˜ â””â”€ front query â”€â”€â”˜ â””â”€ front image ID â”€â”€â”˜ â””â”€ back query â”€â”€â”˜ â””â”€ back image ID â”€â”€â”˜
   *
   * If quantity is not specified, we assume a quantity of 1.
   * Specifying a back query is optional.
   * Specifying an image ID (for each face) is optional.
   *
   * (sorry for jamming this much stuff into one regex ðŸ—¿)
   */

  const trimmedLine = line.replace(/\s+/g, " ").trim();
  const re = new RegExp(
    `^(?:([0-9]*)?[xX]?\\s?(.*?)(?:${SelectedImageSeparator}([A-z0-9_\\-]*))?)?(?:(?:\\s*)${
      "\\" + FaceSeparator
    }(?:\\s*)(.+?)(?:${SelectedImageSeparator}([A-z0-9_\\-]*))?)?$`,
    "gm"
  );
  const results = re.exec(trimmedLine);
  if (results == null) {
    return [0, null, null];
  }
  return [
    parseInt(results[1] ?? "1"),
    [results[2], results[3]],
    [results[4], results[5]],
  ];
}

export function processLine(line: string, dfcPairs: DFCPairs): ProcessedLine {
  /**
   * Process `line` to identify the search query and the number of instances requested for each face.
   * If no back query is specified, attempt to match the front query to a DFC pair.
   * For example, `line`="3x t:goblin" would yield:
   *   [3, {query: {query: "goblin", card_type: TOKEN, selectedImage: null}}, null].
   * Another example is `line`="3x forest | b:elf" would yield:
   *   [
   *     3,
   *     {query: {query: "forest", card_type: CARD}, selectedImage: null},
   *     {query: {query: "elf", card_type: TOKEN}, selectedImage: null},
   *   ].
   */

  const [quantity, frontRawQuery, backRawQuery] = unpackLine(line);

  let frontQuery: SearchQuery | null = null;
  let frontSelectedImage: string | undefined = undefined;
  if (frontRawQuery != null && (frontRawQuery[0] ?? "").length > 0) {
    frontQuery = processPrefix(frontRawQuery[0]);
    frontSelectedImage = frontRawQuery[1] ?? undefined;
  }

  let backQuery: SearchQuery | null = null;
  let backSelectedImage: string | undefined = undefined;
  if (backRawQuery != null && (backRawQuery[0] ?? "").length > 0) {
    backQuery = processPrefix(backRawQuery[0]);
    backSelectedImage = backRawQuery[1] ?? undefined;
  } else if (
    frontQuery != null &&
    frontQuery?.query != null &&
    frontQuery.query in dfcPairs
  ) {
    // attempt to match to DFC pair
    // TODO: is it problematic to assume that all DFC pairs are the `Card` type?
    backQuery = { query: dfcPairs[frontQuery.query], card_type: Card };
  }

  return [
    quantity,
    frontQuery != null
      ? { query: frontQuery, selectedImage: frontSelectedImage }
      : null,
    backQuery != null
      ? { query: backQuery, selectedImage: backSelectedImage }
      : null,
  ];
}

export function processLines(
  lines: string,
  dfcPairs: DFCPairs
): Array<ProcessedLine> {
  /**
   * Process each line in `lines`, ignoring any lines which don't contain relevant information.
   */

  const queries: Array<[number, ProjectMember | null, ProjectMember | null]> =
    [];
  lines.split(/\r?\n|\r|\n/g).forEach((line: string) => {
    if (line != null && line.trim().length > 0) {
      const [quantity, frontMember, backMember] = processLine(line, dfcPairs);
      if (quantity > 0 && (frontMember != null || backMember != null)) {
        queries.push([quantity, frontMember, backMember]);
      }
    }
  });
  return queries;
}

export function standardiseURL(url: string): string {
  /**
   * Standardise `url` in the following ways:
   * 1. Ensure a http prefix is included, defaulting to `https://` if not specified
   * 2. Trim any trailing slash and path
   */

  const re = [...url.matchAll(/^(https?:\/\/)?(.*?)(?:\/.*)?$/gm)][0];
  return (re[1] ?? "https://") + re[2];
}

// TODO: delete this when remaining API interactions have been moved to RTK query
export function formatURL(backendURL: string, routeURL: string): string {
  // TODO: implement this properly
  return backendURL + routeURL;
}
