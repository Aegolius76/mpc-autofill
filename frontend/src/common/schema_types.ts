/* eslint-disable */
/**
 * This file was automatically generated by json-schema-to-typescript.
 * DO NOT MODIFY IT BY HAND. Instead, modify the source JSONSchema file,
 * and run json-schema-to-typescript to regenerate this file.
 */

export interface SearchQuery {
  query: string | null;
  card_type: "CARD" | "CARDBACK" | "TOKEN";
}
/* eslint-disable */
/**
 * This file was automatically generated by json-schema-to-typescript.
 * DO NOT MODIFY IT BY HAND. Instead, modify the source JSONSchema file,
 * and run json-schema-to-typescript to regenerate this file.
 */

/**
 * @minItems 2
 * @maxItems 2
 */
export type SourceRow = [number, boolean];

export interface SearchSettings {
  searchTypeSettings: SearchTypeSettings;
  sourceSettings: SourceSettings;
  filterSettings: FilterSettings;
}
export interface SearchTypeSettings {
  /**
   * Whether fuzzy search is enabled
   */
  fuzzySearch: boolean;
  /**
   * Whether search settings apply to cardbacks or not
   */
  filterCardbacks: boolean;
}
export interface SourceSettings {
  /**
   * The list of sources in the order they should be searched
   */
  sources: SourceRow[] | null;
}
export interface FilterSettings {
  /**
   * The minimum DPI that cards must meet to be included in search results
   */
  minimumDPI: number;
  /**
   * The maximum DPI that cards can have to be included in search results
   */
  maximumDPI: number;
  /**
   * The maximum filesize that cards can have to be included in search results
   */
  maximumSize: number;
}
