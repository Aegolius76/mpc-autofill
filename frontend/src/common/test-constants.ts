/**
 * Some pre-built objects which can be used to build up Redux state for tests.
 */

import { Card, MaximumDPI, MaximumSize, MinimumDPI } from "@/common/constants";
import {
  BackendState,
  CardDocument,
  CardDocumentsState,
  Project,
  SearchResultsState,
  SearchSettings,
  SourceDocument,
  SourceDocuments,
} from "@/common/types";

//# region backend

export const localBackendURL = "https://127.0.0.1:8000";
export const localBackend: BackendState = { url: localBackendURL };
export const noBackend: BackendState = { url: null };

//# endregion

//# region sources

export const sourceDocument1: SourceDocument = {
  pk: 0,
  key: "source_1",
  name: "Source 1",
  identifier: "id_1",
  source_type: "Google Drive",
  external_link: null,
  description: "",
};

export const sourceDocument2: SourceDocument = {
  pk: 1,
  key: "source_2",
  name: "Source 2",
  identifier: "id_2",
  source_type: "Google Drive",
  external_link: null,
  description: "",
};

export const sourceDocument3: SourceDocument = {
  pk: 2,
  key: "source_3",
  name: "Source 3",
  identifier: "id_3",
  source_type: "Google Drive",
  external_link: null,
  description: "",
};

export const sourceDocument4: SourceDocument = {
  pk: 3,
  key: "source_4",
  name: "Source 4",
  identifier: "id_4",
  source_type: "Google Drive",
  external_link: null,
  description: "",
};

export const sourceDocuments: SourceDocuments = {
  [sourceDocument1.pk]: sourceDocument1,
  [sourceDocument2.pk]: sourceDocument2,
  [sourceDocument3.pk]: sourceDocument3,
  [sourceDocument4.pk]: sourceDocument4,
};

//# endregion

//# region cards

export const cardDocument1: CardDocument = {
  identifier: "abc123",
  card_type: "CARD",
  name: "Card 1",
  priority: 0,
  source: sourceDocument1.key,
  source_name: sourceDocument1.name,
  source_id: sourceDocument1.pk,
  source_verbose: sourceDocument1.name,
  source_type: "Google Drive",
  source_external_link: null,
  dpi: 1200,
  searchq: "card 1",
  extension: "png",
  date: "1st January, 2000", // formatted by backend
  download_link: "",
  size: 10_000_000,
  small_thumbnail_url: "",
  medium_thumbnail_url: "",
  language: "EN",
  tags: [],
};

export const cardDocument2: CardDocument = {
  identifier: "abc1234",
  card_type: "CARD",
  name: "Card 2",
  priority: 0,
  source: sourceDocument1.key,
  source_name: sourceDocument1.name,
  source_id: sourceDocument1.pk,
  source_verbose: sourceDocument1.name,
  source_type: "Google Drive",
  source_external_link: null,
  dpi: 1200,
  searchq: "card 2",
  extension: "png",
  date: "1st January, 2000", // formatted by backend
  download_link: "",
  size: 10_000_000,
  small_thumbnail_url: "",
  medium_thumbnail_url: "",
  language: "EN",
  tags: [],
};

export const cardDocument3: CardDocument = {
  identifier: "abc12345",
  card_type: "CARD",
  name: "Card 3",
  priority: 0,
  source: sourceDocument1.key,
  source_name: sourceDocument1.name,
  source_id: sourceDocument1.pk,
  source_verbose: sourceDocument1.name,
  source_type: "Google Drive",
  source_external_link: null,
  dpi: 1200,
  searchq: "card 3",
  extension: "png",
  date: "1st January, 2000", // formatted by backend
  download_link: "",
  size: 10_000_000,
  small_thumbnail_url: "",
  medium_thumbnail_url: "",
  language: "EN",
  tags: [],
};

export const cardDocument4: CardDocument = {
  identifier: "abc123456",
  card_type: "CARD",
  name: "Card 4",
  priority: 0,
  source: sourceDocument1.key,
  source_name: sourceDocument1.name,
  source_id: sourceDocument1.pk,
  source_verbose: sourceDocument1.name,
  source_type: "Google Drive",
  source_external_link: null,
  dpi: 1200,
  searchq: "card 4",
  extension: "png",
  date: "1st January, 2000", // formatted by backend
  download_link: "",
  size: 10_000_000,
  small_thumbnail_url: "",
  medium_thumbnail_url: "",
  language: "EN",
  tags: [],
};

export const cardDocument5: CardDocument = {
  identifier: "abc1234567",
  card_type: "CARDBACK",
  name: "Card 5",
  priority: 0,
  source: sourceDocument2.key,
  source_name: sourceDocument2.name,
  source_id: sourceDocument2.pk,
  source_verbose: `${sourceDocument2.name} Cardbacks`,
  source_type: "Google Drive",
  source_external_link: null,
  dpi: 1200,
  searchq: "card 5",
  extension: "png",
  date: "1st January, 2000", // formatted by backend
  download_link: "",
  size: 10_000_000,
  small_thumbnail_url: "",
  medium_thumbnail_url: "",
  language: "EN",
  tags: [],
};

export const cardDocument6: CardDocument = {
  identifier: "abc12345678",
  card_type: "TOKEN",
  name: "Card 6",
  priority: 0,
  source: sourceDocument3.key,
  source_name: sourceDocument3.name,
  source_id: sourceDocument3.pk,
  source_verbose: `${sourceDocument3.name} Tokens`,
  source_type: "Google Drive",
  source_external_link: null,
  dpi: 1200,
  searchq: "card 6",
  extension: "png",
  date: "1st January, 2000", // formatted by backend
  download_link: "",
  size: 10_000_000,
  small_thumbnail_url: "",
  medium_thumbnail_url: "",
  language: "EN",
  tags: [],
};

//# endregion

//# region project

export const projectSelectedImage1: Project = {
  members: [
    {
      front: {
        query: { query: "my search query", card_type: Card },
        selectedImage: cardDocument1.identifier,
        selected: false,
      },
      back: null,
    },
  ],
  cardback: null,
  mostRecentlySelectedSlot: null,
};

export const projectThreeMembersSelectedImage1: Project = {
  members: [
    {
      front: {
        query: { query: "my search query", card_type: Card },
        selectedImage: cardDocument1.identifier,
        selected: false,
      },
      back: null,
    },
    {
      front: {
        query: { query: "my search query", card_type: Card },
        selectedImage: cardDocument1.identifier,
        selected: false,
      },
      back: null,
    },
    {
      front: {
        query: { query: "my search query", card_type: Card },
        selectedImage: cardDocument1.identifier,
        selected: false,
      },
      back: null,
    },
  ],
  cardback: null,
  mostRecentlySelectedSlot: null,
};

export const projectSelectedImage2: Project = {
  members: [
    {
      front: {
        query: { query: "my search query", card_type: Card },
        selectedImage: cardDocument2.identifier,
        selected: false,
      },
      back: null,
    },
  ],
  cardback: null,
  mostRecentlySelectedSlot: null,
};

//# endregion

//# region search settings

export const defaultSettings: SearchSettings = {
  searchTypeSettings: { fuzzySearch: false, filterCardbacks: false },
  sourceSettings: {
    sources: [
      [sourceDocument1.pk, true],
      [sourceDocument2.pk, true],
      [sourceDocument3.pk, true],
      [sourceDocument4.pk, true],
    ],
  },
  filterSettings: {
    minimumDPI: MinimumDPI,
    maximumDPI: MaximumDPI,
    maximumSize: MaximumSize,
    languages: [],
    includesTags: [],
    excludesTags: ["NSFW"],
  },
};

//# endregion
