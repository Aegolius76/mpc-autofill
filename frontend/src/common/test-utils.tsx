/**
 * This module contains a few utils for testing with Redux and React Testing Library.
 * Retrieved from https://redux.js.org/usage/writing-tests
 */

import type { PreloadedState } from "@reduxjs/toolkit";
import { Store } from "@reduxjs/toolkit";
import { within } from "@testing-library/dom";
import type { RenderOptions } from "@testing-library/react";
import { fireEvent, render, screen, waitFor } from "@testing-library/react";
import FileSaver from "file-saver";
import React, { PropsWithChildren } from "react";
import { Provider } from "react-redux";

import { RootState, setupStore } from "@/app/store";
import { Faces } from "@/common/types";

//# region redux test setup

// This type interface extends the default options for render from RTL, as well
// as allows the user to specify other things such as initialState, store.
interface ExtendedRenderOptions extends Omit<RenderOptions, "queries"> {
  preloadedState?: PreloadedState<RootState>;
  store?: Store;
}

export function renderWithProviders(
  ui: React.ReactElement,
  {
    preloadedState = {},
    // Automatically create a store instance if no store was passed in
    store = setupStore(preloadedState),
    ...renderOptions
  }: ExtendedRenderOptions = {}
) {
  function Wrapper({ children }: PropsWithChildren<{}>): JSX.Element {
    if (store != undefined) {
      return <Provider store={store}>{children}</Provider>;
    } else {
      return <>{children}</>;
    }
  }

  // Return an object with the store and all of RTL's query functions
  // @ts-ignore
  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
}

//# endregion

//# region assertions

export async function expectCardSlotToExist(slot: number) {
  // front and back are both picked up by this query
  await waitFor(() =>
    expect(screen.getAllByText(`Slot ${slot}`)).toHaveLength(2)
  );
}

export async function expectCardSlotToNotExist(slot: number) {
  await waitFor(() =>
    expect(screen.queryAllByText(`Slot ${slot}`)).toHaveLength(0)
  );
}

async function expectCardSlotState(
  testId: string,
  cardName: string | null,
  selectedImage: number | null,
  totalImages: number | null
) {
  /**
   * This function helps with asserting that a particular card slot exists
   * and has the expected image selected.
   */

  await waitFor(() => expect(screen.getByTestId(testId)).toBeInTheDocument());
  const cardElement = screen.getByTestId(testId);
  if (cardName != null) {
    await waitFor(() =>
      expect(within(cardElement).getByText(cardName)).toBeInTheDocument()
    );
  }
  if (selectedImage != null && totalImages != null) {
    await waitFor(() =>
      expect(
        within(cardElement).getByText(`${selectedImage} / ${totalImages}`)
      ).toBeInTheDocument()
    );
  }
}

export async function expectCardGridSlotState(
  slot: number,
  face: Faces,
  cardName: string | null,
  selectedImage: number | null,
  totalImages: number | null
) {
  // note: the specified `slot` should be 1-indexed
  return await expectCardSlotState(
    `${face}-slot${slot - 1}`,
    cardName,
    selectedImage,
    totalImages
  );
}

export async function expectCardbackSlotState(
  cardName: string | null,
  selectedImage: number | null,
  totalImages: number | null
) {
  return await expectCardSlotState(
    "common-cardback",
    cardName,
    selectedImage,
    totalImages
  );
}

//# endregion

//# region react-dropzone

function createDtWithFiles(files: File[] = []) {
  /**
   * Copy/pasted from react-dropzone test source code.
   *
   * createDtWithFiles creates a mock data transfer object that can be used for drop events
   * @param {File[]} files
   */

  return {
    dataTransfer: {
      files,
      items: files.map((file) => ({
        kind: "file",
        size: file.size,
        type: file.type,
        getAsFile: () => file,
      })),
      types: ["Files"],
    },
  };
}

//# endregion

//# region UI interactions

export async function openImportTextModal() {
  // open the modal and find the text area
  screen.getByText("Add Cards", { exact: false }).click();
  await waitFor(() => screen.getByText("Text", { exact: false }).click());
  await waitFor(() => expect(screen.getByText("Add Cards — Text")));
  return screen.getByLabelText("import-text");
}

export async function importText(text: string) {
  const textArea = await openImportTextModal();
  fireEvent.change(textArea, { target: { value: text } });
  screen.getByLabelText("import-text-submit").click();
}

export async function openImportCSVModal() {
  // open the modal and find the upload dropzone
  screen.getByText("Add Cards", { exact: false }).click();
  await waitFor(() => screen.getByText("CSV", { exact: false }).click());
  await waitFor(() => expect(screen.getByText("Add Cards — CSV")));
  return screen.getByLabelText("import-csv");
}

export async function importCSV(fileContents: string) {
  const dropzone = await openImportCSVModal();

  const file = new File([fileContents], "test.csv", { type: "text/csv" });

  fireEvent.drop(dropzone, createDtWithFiles([file]));
}

export async function openImportXMLModal() {
  // open the modal and find the upload dropzone
  screen.getByText("Add Cards", { exact: false }).click();
  await waitFor(() => screen.getByText("XML", { exact: false }).click());
  await waitFor(() => expect(screen.getByText("Add Cards — XML")));
  return screen.getByLabelText("import-xml");
}

export async function importXML(
  fileContents: string,
  useXMLCardback: boolean = false
) {
  const dropzone = await openImportXMLModal();

  const file = new File([fileContents], "test.xml", {
    type: "text/xml;charset=utf-8",
  });
  if (useXMLCardback) {
    await waitFor(() => screen.getByText("Retain Selected Cardback").click());
  }

  fireEvent.drop(dropzone, createDtWithFiles([file]));
}

async function openGridSelector(
  cardSlotTestId: string,
  gridSelectorTestId: string,
  selectedImage: number,
  totalImages: number
) {
  expect(totalImages).toBeGreaterThan(1);
  await expectCardSlotState(cardSlotTestId, null, selectedImage, totalImages);

  await waitFor(() =>
    within(screen.getByTestId(cardSlotTestId))
      .getByText(`${selectedImage} / ${totalImages}`)
      .click()
  );
  await waitFor(() => expect(screen.getByText("Option 1")));

  return screen.getByTestId(gridSelectorTestId);
}

export async function openCardSlotGridSelector(
  slot: number,
  face: Faces,
  selectedImage: number,
  totalImages: number
) {
  return await openGridSelector(
    `${face}-slot${slot - 1}`,
    `${face}-slot${slot - 1}-grid-selector`,
    selectedImage,
    totalImages
  );
}

export async function openSearchSettingsModal() {
  screen.getByText("Search Settings", { exact: false }).click();
  await waitFor(() => expect(screen.getByText("Search Settings")));
  return screen.getByTestId("search-settings");
}

async function downloadFile(id: string) {
  // @ts-ignore
  jest.spyOn(global, "Blob").mockImplementation(function (content, options) {
    return { content, options };
  });
  await waitFor(() => screen.getByTestId(`download-${id}`).click());
  expect(FileSaver.saveAs).toHaveBeenCalledTimes(1);

  // @ts-ignore
  const [blob, filename] = (FileSaver.saveAs as jest.Mock).mock.calls[0];
  return [blob, filename];
}

export async function downloadXML() {
  return await downloadFile("xml");
}

export async function downloadDecklist() {
  return await downloadFile("decklist");
}

//# endregion

//# region misc

export function normaliseString(text: string): string {
  /**
   * Pretty gross, but useful for asserting that generated XMLs are as expected.
   */

  return text.replaceAll(" ", "").replaceAll("\n", "").replaceAll("\r", "");
}

//# endregion
