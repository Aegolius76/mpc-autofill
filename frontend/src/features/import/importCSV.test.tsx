import { screen } from "@testing-library/react";

import App from "@/app/app";
import { cardDocument2, localBackend } from "@/common/test-constants";
import { openImportCSVModal, renderWithProviders } from "@/common/test-utils";

const preloadedState = {
  backend: localBackend,
  project: {
    members: [],
    cardback: cardDocument2.identifier,
  },
};

//# region snapshot tests

test("the html structure of CSV importer", async () => {
  renderWithProviders(<App />, { preloadedState });

  await openImportCSVModal();

  expect(screen.getByTestId("import-csv")).toMatchSnapshot();
});

//# endregion
