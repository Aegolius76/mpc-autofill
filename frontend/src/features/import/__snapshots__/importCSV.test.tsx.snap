// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the html structure of CSV importer 1`] = `
<div
  class="modal-dialog"
  data-testid="import-csv"
>
  <div
    class="modal-content"
  >
    <div
      class="modal-header"
    >
      <div
        class="modal-title h4"
      >
        Add Cards — CSV
      </div>
      <button
        aria-label="Close"
        class="btn-close"
        type="button"
      />
    </div>
    <div
      class="modal-body"
    >
      <p>
        Upload a CSV file of cards to add to the project. The file must
         
        <b>
          exactly
        </b>
         match the following format:
      </p>
      <table
        class="sc-9ba4be8-0 fjcHUO table table-bordered"
      >
        <thead>
          <th
            class="sc-9ba4be8-1 gGMbCh"
          >
            Quantity
          </th>
          <th
            class="sc-9ba4be8-1 gGMbCh"
          >
            Front
          </th>
          <th
            class="sc-9ba4be8-1 gGMbCh"
          >
            Front ID
          </th>
          <th
            class="sc-9ba4be8-1 gGMbCh"
          >
            Back
          </th>
          <th
            class="sc-9ba4be8-1 gGMbCh"
          >
            Back ID
          </th>
        </thead>
        <tbody>
          <tr>
            <td
              class="sc-9ba4be8-2 jEGATG"
            />
            <td
              class="sc-9ba4be8-2 jEGATG"
            />
            <td
              class="sc-9ba4be8-2 jEGATG"
            />
            <td
              class="sc-9ba4be8-2 jEGATG"
            />
            <td
              class="sc-9ba4be8-2 jEGATG"
            />
          </tr>
          <tr>
            <td
              class="sc-9ba4be8-2 jEGATG"
            />
            <td
              class="sc-9ba4be8-2 jEGATG"
            />
            <td
              class="sc-9ba4be8-2 jEGATG"
            />
            <td
              class="sc-9ba4be8-2 jEGATG"
            />
            <td
              class="sc-9ba4be8-2 jEGATG"
            />
          </tr>
          <tr>
            <td
              class="sc-9ba4be8-2 jEGATG"
            />
            <td
              class="sc-9ba4be8-2 jEGATG"
            />
            <td
              class="sc-9ba4be8-2 jEGATG"
            />
            <td
              class="sc-9ba4be8-2 jEGATG"
            />
            <td
              class="sc-9ba4be8-2 jEGATG"
            />
          </tr>
        </tbody>
      </table>
      Where the columns follow these rules:
      <ul>
        <li>
          <b>
            Quantity
          </b>
          : The quantity to include of this row. Must be greater than 0. Can be blank — if blank, a quantity of 1 will be assumed.
        </li>
        <li>
          <b>
            Front
          </b>
          : Search query for card front. 
          <b>
            Cannot be blank.
          </b>
        </li>
        <li>
          <b>
            Front ID
          </b>
          : If this image is in the front search results, it will be pre-selected. Can be blank.
        </li>
        <li>
          <b>
            Front
          </b>
          : Search query for card back. Can be blank.
        </li>
        <li>
          <b>
            Front ID
          </b>
          : If this image is in the back search results, it will be pre-selected. Can be blank.
        </li>
      </ul>
      For example:
      <table
        class="sc-9ba4be8-0 fjcHUO table table-bordered"
      >
        <thead>
          <th
            class="sc-9ba4be8-1 gGMbCh"
          >
            Quantity
          </th>
          <th
            class="sc-9ba4be8-1 gGMbCh"
          >
            Front
          </th>
          <th
            class="sc-9ba4be8-1 gGMbCh"
          >
            Front ID
          </th>
          <th
            class="sc-9ba4be8-1 gGMbCh"
          >
            Back
          </th>
          <th
            class="sc-9ba4be8-1 gGMbCh"
          >
            Back ID
          </th>
        </thead>
        <tbody>
          <tr>
            <td
              class="sc-9ba4be8-2 jEGATG"
            >
              <code>
                2
              </code>
            </td>
            <td
              class="sc-9ba4be8-2 jEGATG"
            >
              <code>
                island
              </code>
            </td>
            <td
              class="sc-9ba4be8-2 jEGATG"
            >
              <code>
                1HsvTYs1...
              </code>
            </td>
            <td
              class="sc-9ba4be8-2 jEGATG"
            >
              <code>
                forest
              </code>
            </td>
            <td
              class="sc-9ba4be8-2 jEGATG"
            />
          </tr>
          <tr>
            <td
              class="sc-9ba4be8-2 jEGATG"
            >
              <code>
                3
              </code>
            </td>
            <td
              class="sc-9ba4be8-2 jEGATG"
            >
              <code>
                t:goblin
              </code>
            </td>
            <td
              class="sc-9ba4be8-2 jEGATG"
            />
            <td
              class="sc-9ba4be8-2 jEGATG"
            />
            <td
              class="sc-9ba4be8-2 jEGATG"
            >
              <code>
                1JtXL6Ca...
              </code>
            </td>
          </tr>
        </tbody>
      </table>
      <hr />
      <div
        class="container"
      >
        <div
          aria-label="import-csv"
          class="sc-d1c6a457-0 iJszHV"
          role="presentation"
          tabindex="0"
        >
          <input
            accept="text/csv,.csv"
            multiple=""
            style="display: none;"
            tabindex="-1"
            type="file"
          />
          Drag and drop a file here, or click to select a file.
        </div>
      </div>
    </div>
    <div
      class="modal-footer"
    >
      <button
        class="btn btn-secondary"
        type="button"
      >
        Close
      </button>
    </div>
  </div>
</div>
`;
